### PR TITLE
Char compare spaces

### PR DESCRIPTION
**Pull request by maciejgrzybek**
CC: @sopel39 

Commits will be squashed.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/type/CharOperators.java [hunk 1]
```diff
@@ -54,39 +55,39 @@ public static boolean notEqual(@SqlType("char(x)") Slice left, @SqlType("char(x)
     @SqlType(StandardTypes.BOOLEAN)
     public static boolean lessThan(@SqlType("char(x)") Slice left, @SqlType("char(x)") Slice right)
     {
-        return left.compareTo(right) < 0;
+        return compareChars(left, right) < 0;
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/type/CharOperators.java:58]: compareChars can return int, not sure if that is acceptable for boolean?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 2]
```diff
@@ -545,6 +546,26 @@ public void testDropTableIfExists()
     }
 
     @Test
+    public void testOrderByChar()
+            throws Exception
+    {
+        assertUpdate("CREATE TABLE char_order_by (c_char char(2))");
+        assertUpdate("INSERT INTO char_order_by (c_char) VALUES(CAST('a' as CHAR(2))), (CAST('a\0' as CHAR(2)))", 2);
```
Comment (akshatnair) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: add another row value that has a value of "a  " (with 2 spaces)

**Review by akshatnair — COMMENTED** [review]
I see this in Travis failure

2016-09-30 10:18:20 SEVERE Error processing Split 20160930_043320_00167_x6uhs.1.1-0  (start = 1475210000333, wall = 4 ms, cpu = 0 ms, calls = 4): NOT_SUPPORTED: No storage type for type: char(2)
2016-09-30 10:18:20 SEVERE Stage 20160930_043320_00167_x6uhs.1 failed
com.facebook.presto.spi.PrestoException: No storage type for type: char(2)
    at com.facebook.presto.raptor.storage.OrcFileWriter.toStorageType(OrcFileWriter.java:327

File: presto-main/src/main/java/com/facebook/presto/type/CharUtils.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import io.airlift.slice.Slice;
+
+public class CharUtils
+{
+    private CharUtils() {}
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/CharUtils.java]: move those methods to Chars.java

File: presto-main/src/main/java/com/facebook/presto/type/CharUtils.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    private static int compareUnsignedBytes(byte a, byte b)
+    {
+        return (a < b) ? -1 : ((a == b) ? 0 : 1);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/CharUtils.java]: from Slice.java:

```
    private static int compareUnsignedBytes(byte thisByte, byte thatByte)
    {
        return unsignedByteToInt(thisByte) - unsignedByteToInt(thatByte);
    }

    private static int unsignedByteToInt(byte thisByte)
    {
        return thisByte & 0xFF;
    }
```

This is easier and faster

File: presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java [hunk 5]
```diff
@@ -90,6 +100,11 @@ public void testGreaterThan()
         assertFunction("cast('bar' as char(3)) > cast('foo' as char(3))", BOOLEAN, false);
         assertFunction("cast('foobar' as char(6)) > cast('foobaz' as char(6))", BOOLEAN, false);
         assertFunction("cast('foob r' as char(6)) > cast('foobar' as char(6))", BOOLEAN, false);
+        assertFunction("cast(' ' as char(1)) > cast('\0' as char(1))", BOOLEAN, true);
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java:111]: keep the same arguments order as in other tests?

File: presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java [hunk 6]
```diff
@@ -56,6 +56,11 @@ public void testLessThan()
         assertFunction("cast('bar' as char(3)) < cast('foo' as char(3))", BOOLEAN, true);
         assertFunction("cast('foobar' as char(6)) < cast('foobaz' as char(6))", BOOLEAN, true);
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java:63]: add test case:
`assertFunction("cast('\0 a' as char(3)) > cast('\0' as char(3))", BOOLEAN, TRUE)`
`assertFunction("cast('\0 a' as char(3)) < cast('\0' as char(3))", BOOLEAN, FALSE)`

File: presto-main/src/main/java/com/facebook/presto/type/CharUtils.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        if (left.length() < right.length()) {
+            return charCmp(left, right);
+        }
+        else {
+            return -charCmp(right, left);
+        }
+    }
+
+    private static int charCmp(Slice shorter, Slice longer)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/CharUtils.java]: rename to: `comapreCharsShorterToLonger`?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 8]
```diff
@@ -545,6 +546,26 @@ public void testDropTableIfExists()
     }
 
     @Test
+    public void testOrderByChar()
+            throws Exception
+    {
+        assertUpdate("CREATE TABLE char_order_by (c_char char(2))");
```
Comment (sopel39) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: not all connectors support char (e.g raptor). Maybe remote this test or move it to: `TestHiveDistributedQueries`?

**Comment by sopel39** [comment]
add test line to: `TestTupleDomainOrcPredicate#testChar()`:
I think:
`assertEquals(getDomain(CHAR, 10, stringColumnStats(10L, "\0         ", "          ")), create(ValueSet.ofRanges(range(CHAR, utf8Slice("\0"), true, utf8Slice(""), true)), false));`
will do

**Comment by sopel39** [comment]
comments & LGTM

**Comment by sopel39** [comment]
@KBP-TDC Should this be backported to 148t? How do we proceed?

**Comment by ilfrin** [comment]
this needs to be backported to 148t

File: presto-main/src/main/java/com/facebook/presto/type/CharOperators.java (hunk 1 above)
Comment (maciejgrzybek) [reply presto-main/src/main/java/com/facebook/presto/type/CharOperators.java:58]: That's why there's `< 0`.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java (hunk 2 above)
Comment (maciejgrzybek) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Sure. It does not test comparison, though. It'd test truncation of longer than defined chars.

File: presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java (hunk 5 above)
Comment (maciejgrzybek) [reply presto-main/src/test/java/com/facebook/presto/type/TestCharOperators.java:111]: '\0' is < than ' '. There is no letter < than ' ', that's why this test is different than others thus important

**Comment by maciejgrzybek** [comment]
Closing in favor of prestodb#6233

**Pull request closed without merging**